### ISSUE TITLE
Totality/positivity checker issue

### ISSUE THREAD

**Issue by mb64** [idris-lang/Idris2#524]
It's possible to prove `Void` in a `total` way using non-positive data:
```idris
$ cat OhNo.idr 
module OhNo

data Bad = MkBad (Not Bad)

hmmm : Bad -> Not Bad
hmmm (MkBad n) = n

ok : Not Bad
ok bad = hmmm bad bad

bad : Bad
bad = MkBad ok

total
ohno : Void
ohno = ok bad
$ idris2 --no-banner -q OhNo.idr 
OhNo> :t ohno
OhNo.ohno : Void
OhNo> :total ohno
OhNo.ohno is total
```

I expect this to not compile, and that `hmmm`, `ok`, `bad`, and `ohno` should be marked as not total since they use `Bad`, which it correctly identifies as not strictly positive.

This seems to be a regression from Idris 1, which correctly flags all relevant functions as not total.

**Comment by donovancrichton** [comment]
There seems to be some different behaviour happening between `total` and `%default total`
If I run the following:
```Idris
%default total
data Bad = MkBad (Not Bad)

hmmm : Bad -> Not Bad 
hmmm (MkBad n) = n 

ok : Not Bad 
ok bad = hmmm bad bad 

bad : Bad 
bad = MkBad ok

ohno : Void
ohno = ok bad 
```
I correctly get:
```Idris
1/1: Building bad (bad.idr)
Error: Bad is not total, not strictly positive

bad.idr:2:1--4:5
 2 | data Bad = MkBad (Not Bad)
 3 | 
 4 | hmmm : Bad -> Not Bad

Error: MkBad is not total, not strictly positive

bad.idr:2:12--2:27
   |
 2 | data Bad = MkBad (Not Bad)
   |            ^^^^^^^^^^^^^^^

Error: bad is not total, possibly not terminating due to call to Main.MkBad

bad.idr:10:1--11:4
 10 | bad : Bad
 11 | bad = MkBad ok

Error: ohno is not total, possibly not terminating due to call to Main.bad

bad.idr:13:1--14:5
 13 | ohno : Void
 14 | ohno = ok bad
```

**Comment by gallais** [comment]
I think I have found the source of this. When we are checking some
requirement, we skip an `Unchecked` (presumably because we assumed
it was impossible because definitions would be checked as they are
defined?) rather than checking it.

**Comment by gallais** [comment]
I think the problem comes from the fact we ignore [this `Unchecked`](https://github.com/idris-lang/Idris2/blob/a7cb2745bcc7cf6b920ff00eecb18372835c4a06/src/Core/Termination.idr#L28). Unfortunately
it's not as simple as replacing this with a `checkTotal` because that would lead
to an infinite loop.

I have also tried to add a mandatory positivity check [here](https://github.com/idris-lang/Idris2/blob/a7cb2745bcc7cf6b920ff00eecb18372835c4a06/src/TTImp/ProcessDecls.idr#L71) to fix the fact this was
overlooked in case `total` is not turned on. But, unlike the termination checker,
AFAICT the check making sure we are not referencing other non-total functions
does not follow the dependency graph. Meaning that we may hit the `Unchecked`
issue above.

I have also tried to `checkTotal` no matter what and only raise errors if we are
in the `total` case but that slows the compiler down *dramatically* and I can't get
the compilation past `Core.Hash` in a reasonable amount of time.

Thoughts @edwinb?

**Comment by clayrat** [comment]
@gallais Does https://github.com/idris-lang/Idris2/commit/c3082d8465f42345842aa36d3d7993d6b2f76712 solve this?

**Comment by gallais** [comment]
Nope. It fixes the fact that `Bad` was not recorded as not strictly positive
but we still have the problem that the totality of `ohno` is checked before
that of `ok` and `bad` and thus we only see an `Unchecked` that we ignore
[here](https://github.com/idris-lang/Idris2/blob/a7cb2745bcc7cf6b920ff00eecb18372835c4a06/src/Core/Termination.idr#L28) when checking that all of the definitions mentioned in `ohno` are safe.

**Comment by dunhamsteve** [comment]
This was fixed by #3043.  It looks like I mistyped this issue number in the PR description but correctly named the test file `Issue-524.idr`.
